master/buildbot/reporters/notifier.py

```diff
@@ -55,85 +48,104 @@ def checkConfig(self, mode=("failing", "passing", "warnings"),
-        if builders is not None and tags is not None:
-            config.error(
-                "Please specify only builders or tags to include - " +
-                "not both.")
-
-        if not(watchedWorkers == 'all' or watchedWorkers is None or
-                isinstance(watchedWorkers, (list, tuple, set))):
-            config.error("watchedWorkers must be 'all', None, or list of worker names")
+            for g in generators:
+                self.name += "_" + g.generate_name()
 
     def reconfigService(self, mode=("failing", "passing", "warnings"),
                         tags=None, builders=None,
                         buildSetSummary=False, messageFormatter=None,
                         subject="Buildbot %(result)s in %(title)s on %(builder)s",
                         addLogs=False, addPatch=False,
                         schedulers=None, branches=None,
-                        watchedWorkers=None, messageFormatterMissingWorker=None):
-
-        self.mode = self.computeShortcutModes(mode)
-        self.tags = tags
-        self.builders = builders
-        self.schedulers = schedulers
-        self.branches = branches
-        self.subject = subject
-        self.addLogs = addLogs
-        self.addPatch = addPatch
-        if messageFormatter is None:
-            messageFormatter = DefaultMessageFormatter()
-        self.messageFormatter = messageFormatter
-        if messageFormatterMissingWorker is None:
-            messageFormatterMissingWorker = MessageFormatterMissingWorker()
-        self.messageFormatterMissingWorker = messageFormatterMissingWorker
-        self.buildSetSummary = buildSetSummary
-        self._buildset_complete_consumer = None
-        if watchedWorkers is None:
-            self.watchedWorkers = ()
-        else:
-            self.watchedWorkers = watchedWorkers
+                        watchedWorkers=None, messageFormatterMissingWorker=None,
+                        generators=None):
+
+        if generators is None:
+            generators = self.create_generators_from_old_args(mode, tags, builders, buildSetSummary,
+                                                              messageFormatter, subject, addLogs,
+                                                              addPatch, schedulers,
+                                                              branches, watchedWorkers,
+                                                              messageFormatterMissingWorker)
+
+        self.generators = generators
+        self._needs_buildsets_complete = \
+            any(isinstance(g, BuildSetStatusGenerator) for g in generators)
+        self._needs_build_finished = \
+            any(isinstance(g, BuildStatusGenerator) for g in generators)
+        self._needs_worker_missing = \
+            any(isinstance(g, WorkerMissingGenerator) for g in generators)
 
     @defer.inlineCallbacks
     def startService(self):
         yield super().startService()
```

Comment:
this calls reconfigureService() eventually

Reply:
Fixed now.